File: 0020/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        opens = []
        for char in s:
            if char == "(" or char == "{" or char == "[":
                opens.append(char)

            else:
                if char == "}":
                    if not opens or opens.pop() != "{":
                        return False
                if char == "]":
                    if not opens or opens.pop() != "[":
                        return False
                if char == ")":
                    if not opens or opens.pop() != "(":
                        return False

        return True if opens == [] else False

        # timeO(n)
        # space O(n)

File: 0020/solution.java
import java.util.Map;
import java.util.HashMap;

class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> parens = new HashMap<>();
        parens.put('(', ')');
        parens.put('[', ']');
        parens.put('{', '}');
        StringBuilder builder = new StringBuilder();
        for (char c : s.toCharArray()) {
            if ("({[".indexOf(c) != -1) {
                builder.append(c);
            } else if (!builder.isEmpty() && parens.get(builder.charAt(builder.length() - 1)) == c) {
                builder.deleteCharAt(builder.length() - 1);
            } else {
                return false;
            }
        }
        return builder.isEmpty();
    }
}